src/main/java/co/cask/hydrator/salesforce/SalesforceSchemaUtil.java

```diff
@@ -0,0 +1,149 @@
+
+    String sObjectName = SalesforceQueryParser.getSObjectFromQuery(query);
+    Map<String, Field> nameToField = getAllSObjectFieldsMap(credentials, sObjectName);
+
+    List<String> queryFields = SalesforceQueryParser.getFieldsFromQuery(query);
+    return getSchemaWithFields(queryFields, nameToField);
+  }
+
+  private static Map<String, Field> getAllSObjectFieldsMap(AuthenticatorCredentials credentials, String sObjectName)
+    throws ConnectionException {
+    ConnectorConfig connectorConfig = Authenticator.createConnectorConfig(credentials);
+
+    PartnerConnection partnerConnection = new PartnerConnection(connectorConfig);
+    DescribeSObjectResult describeSObjectResult = partnerConnection.describeSObject(sObjectName);
+
+    Map<String, Field> nameToField = new HashMap<>();
+    for (Field field : describeSObjectResult.getFields()) {
+      nameToField.put(field.getName(), field);
+    }
+    return nameToField;
+  }
+
+  @VisibleForTesting
+  static Schema getSchemaWithFields(List<String> queryFields, Map<String, Field> nameToField) {
+
+    List<Schema.Field> schemaFields = new ArrayList<>(queryFields.size());
+    for (String fieldName : queryFields) {
+      Field salesforceField = nameToField.get(fieldName);
+      Schema fieldSchema;
+
+      if (isFieldComplex(fieldName)) {
+        // Some complex functions like nested and aggregate are not supported by Bulk API
+        // for relationship fields it would take too long to retrieve all fields for every sObject.
+        // Describe calls are a bit slow.
+        fieldSchema = Schema.nullableOf(Schema.of(Schema.Type.STRING));
+      } else {
+        if (salesforceField == null) {
+          throw new IllegalArgumentException(String.format("Field '%s' is absent the sObject", fieldName));
+        }
+
+        FieldType fieldType = salesforceField.getType();
+        Schema.Type schemaFieldType = fieldTypeToCdapSchemaType(fieldType);
+
+        if (salesforceField.isNillable()) {
+          fieldSchema = Schema.nullableOf(Schema.of(schemaFieldType));
+        } else {
+          fieldSchema = Schema.of(schemaFieldType);
+        }
+      }
+
+      Schema.Field schemaField = Schema.Field.of(fieldName, fieldSchema);
+      schemaFields.add(schemaField);
+    }
+
+    return Schema.recordOf("output",  schemaFields);
+  }
+
+  /**
+   * If this is relationship query (join equivalent for soql)
+   * or this is aggregate function, or nested query as field,
+   * don't convert it from string to anything as this is very error-prone.
+   *
+   * Examples of complex fields:
+   * Relationship field - Account.Name
+   * Nested field - (SELECT Id FROM Opportunity)
+   * Aggregate field - AVG(Amount), COUNT()
+   *
+   *
+   * @param fieldName name of the field
+   * @return is field complex (relationship or aggregate or nested)
+   */
+  private static boolean isFieldComplex(String fieldName) {
+    return (fieldName.contains("(") || fieldName.contains("."));
```

Comment:
is it possible to have something like (SELECT Id FROM Opportunity) without it being nested?

Reply:
Salesforce Bulk API responds with an error to "(SELECT Id FROM Opportunity)" query.
> BulkAPIBatchException: Batch failed. BatchId='7511i000002aj0uAAA', Reason='InvalidBatch : Failed to process query: MALFORMED_QUERY: unexpected token: ('